### ISSUE TITLE
Resource versions not being passed

### ISSUE THREAD

**Issue by jmcarp** [concourse/concourse#1298]
# Bug Report

We recently noticed that resource versions are never passed from one task to the next for some pipelines. Given two jobs like this:

```yaml
- name: job1
  - get: resource1
    trigger: true
- name: job2
  - get: resource1
    passed: [job1]
    trigger: true
```

The second job does not trigger when `resource1` updates, and doesn't find versions of `resource1` when triggered manually. I haven't been able to replicate this with a simple example pipeline, and I found that creating a new pipeline with exactly the same configuration as the broken pipeline works as expected, so I'm guessing this is related to something in the pipeline history. Has anybody else experienced this kind of issue, or have tips for debugging?

* Concourse version: 3.1.1/3.2.0
* Deployment type (BOSH/Docker/binary): BOSH
* Infrastructure/IaaS: AWS
* Did this used to work? Yes

cc @LinuxBozo @cnelson

**Comment by vito** [comment]
I'm confused by the example. Do you mean "job" rather than "task"?

Can't think of anything that would change how this works. If you find a concrete example, in lieu of a repro, could you post more evidence, e.g. the version history of the resource, and information around which versions have passed through which jobs?

**Comment by jmcarp** [comment]
<img width="899" alt="screen shot 2017-06-16 at 11 08 48 am" src="https://user-images.githubusercontent.com/1633460/27232560-7c45b6ae-5284-11e7-924e-345b09d01713.png">

<img width="1118" alt="screen shot 2017-06-16 at 11 08 59 am" src="https://user-images.githubusercontent.com/1633460/27232571-88232e66-5284-11e7-82a4-8f32ec123609.png">

Here are some pictures: the `common-development` resource is supposed to be passed from `deploy-cf-development` to `deploy-diego-development`. When I update that resource, `deploy-cf-deployment` triggers, but after that job finishes, `deploy-diego-development` doesn't trigger. Manually triggering `deploy-diego-development` doesn't use the updated version of `common-development`.

Meanwhile, pausing the original pipeline and creating a second pipeline with exactly the same configuration but no build history works as expected.

**Comment by wernerb** [comment]
Is common-development a git resource? do you have paths set? if so make sure that the commits you do target those paths, otherwise concourse will not fetch a new version of the common-development.  What happens with a new pipeline is that concourse has NO version of common-development, so it will retrieve *any* version it will get and so will work for you the first time.

**Comment by LinuxBozo** [comment]
common-development is not a git resource. It is a custom resource that polls a set of paths from S3, and returns as it's version the epoch timestamp of the newest file. However, even if we were running into the situation that you are trying to describe, we would not get the first job to trigger either, and it does so happily when any of the files are updated, so that is not the issue.

**Comment by cnelson** [comment]
We have also observed this happening with the built-in `bosh-io-release`.  When trying to debug this we ran into https://github.com/concourse/concourse/issues/1096 which prevented us from gathering any more useful information.

**Comment by cnelson** [comment]
@vito any more thoughts on how to debug this before we destroy this pipeline and lose the ability to reproduce?

**Comment by cnelson** [comment]
@chendrix @vito  More updates on this.  We've now seen Concourse fail to grab the correct version of the resource even when there are are no passed constraints:

We have this job which pulls in a certificate from a versioned s3 resource (master-bosh-root-cert):

<img width="539" alt="screen shot 2017-06-24 at 16 56 35" src="https://user-images.githubusercontent.com/604163/27512765-71bab7f6-5901-11e7-84df-8eb7d975edb1.png">

Today when attempting to deploy it was grabbing an old version (`SZD...`), there are no passed constraints or anything else involved, this is the first job in a pipeline pulling a file from s3:

<img width="1265" alt="screen shot 2017-06-24 at 16 56 25" src="https://user-images.githubusercontent.com/604163/27512776-ba3d322e-5901-11e7-912c-224dc06b39fd.png">

Examining the resource shows a newer version available (`kaU...`) but it's not used.  Other pipelines that rely on this same resource were pulling the correct version:

<img width="1276" alt="screen shot 2017-06-24 at 16 56 44" src="https://user-images.githubusercontent.com/604163/27512780-d088601c-5901-11e7-8993-38690e3bfdac.png">

If I rename the resource to something else and refly the pipeline it grabbed the correct version.  If I then rename the resource back to it's original name, the pipeline reverted to grabbing the wrong version.

It's really concerning that we can't trust concourse to be pulling in the correct versions.  In this case it was a cert which being out of date caused a deploy to fail so we immediately noticed it, but if it was a stemcell or other security update that wouldn't break a build, and it's not getting pulled in, that's a big deal.

Any ideas on how to debug this?

**Comment by chendrix** [comment]
Is there a version in your pipeline yml that pins the version that's get?

**Comment by cnelson** [comment]
There's no version pinning involved.  No params are being passed to the get:
```
jobs:
- name: deploy-shibboleth-staging
  plan:
  - aggregate:
    - get: master-bosh-root-cert
    - get: pipeline-tasks
    - get: shibboleth-config
      trigger: true
    - get: common
      trigger: true
      resource: common-staging
    - get: terraform-yaml
      resource: terraform-yaml-staging
    - get: cg-s3-shibboleth-release
      trigger: true
    - get: shibboleth-stemcell
      trigger: true
```

**Comment by dsabeti** [comment]
Howdy all. The CF Release Integration team has two pipelines showing this bug as well. It's new in the last few days.

We're running Concourse 3.1.1 on GCP. @vito @chendrix, let me know what other information I can provide.

**Comment by chendrix** [comment]
Looks like the pipelines affected are:

18F: https://github.com/18F/cg-deploy-shibboleth/blob/master/ci/pipeline.yml
RelInt: https://github.com/cloudfoundry/runtime-ci/blob/master/pipelines/cf-release.yml

We've been given version dumps and @vito and I are looking at them in the context of the build algorithm.

**Comment by chendrix** [comment]
From @dsabeti

> Hey, FWIW, after letting a pipeline run go through from beginning to end, with no changes to any of the resources, I got one of the downstream job to pick up the correct resource version

This tends to support the idea that at least for the RelInt case it was a complicated `passed` constraint that wasn't being satisfied.

**Comment by dsabeti** [comment]
Yeah. I'm trying the same approach to fixing the other pipeline that's seeing this issue. I'll report back here once it's finished.

**Comment by vito** [comment]
@cnelson I've added a test [here](https://github.com/concourse/atc/commit/728a43bec4913655997d1efd7d12ddb5d58bc8f3) locally to reflect your job's configuration running against the data set, but the data set itself hows that the job has run with the latest input version (id 206635 of resource 893, `master-bosh-root-cert`). The test result also shows it selecting that as the candidate version.

Do you have a pipeline that's currently in this broken state? Does that result not reflect your expectation for which version should be chosen?

You can get the version IDs by navigating to an API endpoint like the following: https://ci.concourse.ci/api/v1/teams/main/pipelines/main/resources/concourse/versions

The test needs an `18f.json.gz` file under `db/algorhtm/testdata` in the ATC. You could try reproducing this yourself using the `versions-db` API endpoint plus the IDs you expect the job to get based on browsing around the API.

**Comment by chendrix** [comment]
@jmcarp told me that @cnelson was out for the day, but maybe @jmcarp can look at it.

**Comment by chendrix** [comment]
@dsabeti also said he triggered a new build early on that then maybe "flushed" the system's resources and now it's got the correct version

**Comment by andrewedstrom** [comment]
My understanding is that this is currently blocked on needing responses

**Comment by topherbullock** [comment]
Some paths for investigating this: 
- Use the versions from a known failing pipeline + ATC (there's one in the Concourse team's Google drive)
- Write a test in the `db/algorithm` package that fails according to the dataset provided
- Pair with @vito to write the test and fix it

**Comment by chendrix** [comment]
Added the information needed to debug to [Google Drive](https://drive.google.com/drive/folders/0Bzkbg8faXyA3b2ZxQWM3bUdicGs)

**Comment by clarafu** [comment]
The information in the drive isn't enough; we still need to know what the expected versions are in order to write a test.

/cc @cnelson @dsabeti

**Comment by chendrix** [comment]
I'd reach out to @jmcarp then

**Comment by cnelson** [comment]
We still have @jmcarp's `common-development` related pipeline in this state. To reiterate the configuration:

```yaml
- name: deploy-diego-development
  serial: true
  serial_groups:
  - development
  plan:
  - aggregate:
    - get: common-development
      passed:
      - deploy-cf-development
...
```

When starting the job it is using version `1496779473` of the resource:

```json
{
"id":144630,
"team_name":"main",
"name":"71",
"status":"aborted",
"job_name":"deploy-diego-development",
"url":"/teams/main/pipelines/deploy-cf/jobs/deploy-diego-development/builds/71",
"api_url":"/api/v1/builds/144630",
"pipeline_name":"deploy-cf",
"start_time":1502466147,
"end_time":1502466157
}
```

`/api/v1/builds/144630/resources | jq '.inputs[] | select(.name == "common-development") | .version`
```json
{
  "ref": "1496779473"
}
```
Even though we would expect it to use  `1497562788` as that version has successfully passed `deploy-cf-development` much earlier (see timestamps)

`/api/v1/teams/main/pipelines/deploy-cf/resources/commondevelopment/versions/153588/output_of`
```json
[{
"id":85609,
"team_name":"main",
"name":"99",
"status":"succeeded",
"job_name":"deploy-cf-development",
"url":"/teams/main/pipelines/deploy-cf/jobs/deploy-cf-development/builds/99",
"api_url":"/api/v1/builds/85609",
"pipeline_name":"deploy-cf",
"start_time":1497563088,
"end_time":1497563296
}]
```

Here is the  output of `/api/v1/teams/main/pipelines/deploy-cf/resources/common-development/versions` as requested: https://gist.github.com/cnelson/72cfb9ceb7b40fed9e742ea1d3986413

**Comment by vito** [comment]
@cnelson Sorry for the slow feedback loop; never saw your comment.

I'm not sure if you still have the data, but we would need the `id` field, not the `version`, to be able to have a test case.

Something like:

```
/api/v1/teams/main/pipelines/deploy-cf/resources/common-development/versions
```

...and then getting the `id` attributes for the ones pertaining to ref `1496779473` and `1497562788`. It looks like one of them has ID `153588`, from your second snippet?

**Comment by chris-griffin** [comment]
I believe we are running into the same issue. We have a deploy job that we manually trigger using the web console. One of the inputs is the output of another job. That job will show as completed in the web console, but the deploy job will pull the previous version.

It seems like there is some type of cache/update issue. As an aside, this occurs more frequently when concourse is under heavy load. Also, the resource won't show as being "outputs of"  or as being an "input to" the deploy job while we are running into this issue which typically lasts for a few minutes (see screenshot).

![image](https://user-images.githubusercontent.com/1795659/35891951-d67005e0-0b74-11e8-8d5b-34940f126ed3.png)

I can replicate this fairly often so if you have specific debug info you need, please let me know.

**Comment by topherbullock** [comment]
It seems possible for a pending build of a job to "skip" versions of a resource, due to some regression in the scheduling and locks around build tracking. Some reports have been pointing to some combination of a job being `serial:true` or in one of the pipeline's `serial_groups` and using `aggregate` within the build plan.

Some users have noted issues with a pipeline as simple as:
```
- name: job
  serial: true
  plan:
  - aggregate:
    - get: some-resource
    - get: another-resource    
  - task: some-task
   ... 
```
Setting this pipeline with `serial:true`  and triggering a new build will occasionally result in the build sitting in the pending state (seemingly) indefinitely. Removing `serial:true` seems to resolve the lack of versions available, as does restarting the ATCs while the build is stuck in pending state.

**Comment by chris-griffin** [comment]
As @phillbaker pointed out it appears that `REFRESH MATERIALIZED VIEW CONCURRENTLY` added in https://github.com/concourse/concourse/issues/1589 is causing some issues for write heavy (frequent build) environments.

The problem is that only one `REFRESH MATERIALIZED VIEW CONCURRENTLY` can occur at a time:

```
CONCURRENTLY Refresh the materialized view without locking out concurrent selects on the materialized view. (...)

... OTHER CONTENTS ...

Even with this option only one REFRESH at a time may run against any one materialized view.
```

So with frequent builds, the `REFRESH MATERIALIZED VIEW CONCURRENTLY transition_builds_per_job` refresh begin to queue.

This [gets triggered as the last step during the `Finish()` function](https://github.com/concourse/atc/blob/2934cec766645a97f4067e456fa7ee784490c76f/db/build.go#L357). 

**Mitigation idea**

- Only execute the `REFRESH MATERIALIZED VIEW CONCURRENTLY` if one is not already executing or in the queue for that particular mat view

   Something like this will still be susceptible to race conditions, but at least it will prevent more than 1 or 2 getting queue'd (we often see this number > 25)

    ```sql
         select count(*) from pg_stat_activity where state != 'idle' and query = 'REFRESH MATERIALIZED VIEW CONCURRENTLY transition_builds_per_job';
    ```

**Comment by jacob-hd** [comment]
I believe we are experiencing this issue as well.

I've documented our experience so far here:
https://github.com/quotecenter/concourse-issue

I'll continue to update this repository with information as we progress.

**Comment by vito** [comment]
Could this be https://github.com/concourse/concourse/issues/2895?

**Comment by phillbaker** [comment]
@vito looks very similar!

**Comment by phillbaker** [comment]
Would ask that this issue and/or https://github.com/concourse/concourse/issues/736 remain open as it's a common problem and the proposed fix (spaces) is a large and uncertain features (https://github.com/concourse/rfcs/pull/24).

**Comment by vito** [comment]
I'll leave #736 open as long as the issue still stands. The resolution will probably be deprecating/disallowing `version: every` + `passed:` in favor of spatial pipelines. The problem with `version: every` + `passed:` is pretty fundamental as noted in https://github.com/concourse/concourse/issues/736#issuecomment-479557411, and supporting it adds a *lot* of complexity in our scheduling algorithm.

I'm going to close this issue though. It looks like a few people chimed in to say they're also affected, but it's been a couple years and we never found a root cause aside from #2895, which only affected `version: every` and was fixed in v5.1.

It's a little unsatisfying to close this without 100% resolution, but I think it'd be best if new issues are opened with a heavy focus on being able to reproduce it or at least convey a lot of information. If the issue is still there at all. :thinking:

There's also a chance that the new algorithm (#3602) gets rid of any spooky ghost bugs in the algorithm. It at least feels a lot more technically sound. If anyone wants to read about it, there's more information in [this part of the v10 roadmap update](https://blog.concourse-ci.org/core-roadmap-towards-v10/#where-are-we-now).

Thanks all!
